### PR TITLE
Store-behaviour : Adding current state to merging

### PR DESCRIPTION
**Pull request by Hartorn**
Linked to #1477

### CODE REVIEW

**Comment by Hartorn** [comment]
@c3dr0x Basically, if we dont filter default data and merge with current state, it should work, no ?

File: src/common/mixin/store-behaviour.js [hunk 1]
```diff
@@ -42,39 +42,41 @@ const storeMixin = {
...
     /**
      * Get the state informations from the store.
-     * @param  {array} filterNodes - An object containing nodes key to update.
+     * @param  {array} filterNodesArg - An object containing nodes key to update.
      * @returns {object} - The js object constructed from store data.
      */
-    _getStateFromStores(filterNodes = []) {
+    _getStateFromStores(filterNodesArg) {
         if (this.getStateFromStore) {
             return this.getStateFromStore();
```
Comment (c3dr0x) [inline src/common/mixin/store-behaviour.js]: I wonder if we shouldn't push `filterNodesArg` to the override method `getStateFromStore`

**Comment by c3dr0x** [comment]
I thought about it again, and. At 1st iteration it will work as expected. But from from the 2nd and on, wet keep the issue of default values losing in favor of the ones from the state, so KO

iteration1:
- node: `{ a: value1, b: value 2 }`
- state: `null`
- default: `{ a: null, b: null, c: null }`
=> `{ a: value1, b: value2, c: null  }`

iteration2:
- node: `{ a: value3 }`
- state: `{ a: value1, b: value2, c: null  }`
- default: `{ a: null, b: null, c: null }`
=> `{ a: value3, b: value2, c: null  }`

we kept { b: value2 }, and that is not what's intended... 

I think the solution might be adding a new value `definitionInNode` (or something like that) to determine in which node is the entity mapped by the definition.

**Comment by Hartorn** [comment]
I agree with you, but if you need both `definitionInNode` and `useDefaultStoreData `, it's starting to be a lot, for something that should be native.

I went back to something safe, so at least `_getStateFromStores` is sending data with default values.
I think we should get rid of `useDefaultStoreData `, I don't see case where this should break something.

Overall, the best thing would be to define default data on store, on a by node base, either directly or by giving a definitionPath. It is not worth so much time, worst case server should send data with property at null. 

We can add an option, but no one will use it, I think this is a dead end.

**Comment by c3dr0x** [comment]
So three main cases:

1- no `getDefaultStoreData`, no `definitionInNode` => simply retrieves node
2- `getDefaultStoreData` present (with node level key) => retrieves node + merge with equivalent key from `getDefaultStoreData`
3- `definitionInNode` present => retrieves node + if same node as def node, add default null values from definition

I think : 2 & 3 can be cumulative with 2 winning over 3.

Remove `useDefaultStoreData` in general: `getDefaultStoreData` and/or `definitionInNode` are enough

And maybe deport 2 & 3 configuration into store. But I think it should be in addition. It has advantages: extra conf for node kept in store are therefore available to all components. 
But in Atlas we use a lot `getDefaultStoreData` and I don't want to lose it.

**Comment by Hartorn** [comment]
That should do it.

**Comment by Hartorn** [comment]
@c3dr0x Looks good to me.
I let you decide when to merge it, I was the creator of the PR.

**Comment by c3dr0x** [comment]
@harton I’m just waiting to move getState from willMount to didMount as discussed yesterday

**Comment by Hartorn** [comment]
Still not me, I'm Hartorn

Le 18 oct. 2017 8:47 PM, "Cédric BOULET KESSLER" <notifications@github.com>
a écrit :

@Harton <https://github.com/harton> I’m just waiting to move getState from
willMount to didMount as discussed yesterday

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub
<https://github.com/KleeGroup/focus-components/pull/1479#issuecomment-337690545>,
or mute the thread

**Comment by Hartorn** [comment]
@c3dr0x Should we go with that ?

**Comment by c3dr0x** [comment]
Yep, I just tried it, it's okay for me !

**Pull request merged**